ichnaea/api/views.py

```diff
@@ -120,6 +121,23 @@ def preprocess_request(self):
             # Use str(), since repr() includes the full source bytes
             raise self.prepare_exception(ParseError({"decode": str(exc)}))
 
+        if self.ip_log_and_rate_limit:
+            # Count unique daily requests
+            siggen = sha512(content.encode())
+            client_addr = self.request.client_addr or "127.0.0.1"
+            siggen.update(client_addr.encode())
+            siggen.update(self.request.url.encode())  # Includes the API key if given
+            signature = siggen.hexdigest()
+            today = util.utcnow().date().isoformat()
+            key = f"apirequest:{today}"
+            with self.redis_client.pipeline() as pipe:
+                pipe.pfadd(key, signature)
+                pipe.expire(key, 90000)  # 25 hours
+                new_request, _ = pipe.execute()
+            bind_threadlocal(
+                api_repeat_request=not new_request, api_request_sig=signature[:8]
+            )
```

Comment:
The full signature is used with ``PFADD``, so we get the cardinality estimate of HyperLogLog (standard error of 0.81%) for ``api_repeat_request=True``/``False``. The signature prefix is used to count duplicates, and I was more comfortable storing 8 characters than 128. I might be pre-optimizing. I was thinking of two cases.

If we start recording unique requests to ``/v1/country``, no users send radio data, so we'd be hashing ``{}`` + the user IP + the request URL, meaning there might be enough data to figure out the IP from the signature. Only passing the first 8 may make it harder. But that's still 32 bits of data, which might be enough for IPv4.

Second, I was wondering if we could reduce storage needs by storing less of the signature. A full signature is 128 hex digits, but I was hoping for an insignificant number of collisions with 8 digits and only looking at records with``api_repeat_request=True``

The ratio of ``api_repeat_request=True`` to ``api_repeat_request=False`` gives most of the data for [bug 1624687](https://bugzilla.mozilla.org/show_bug.cgi?id=1624687). If 80% are repeated, there is an opportunity for caching in a proxy or in the client, and if 10% are repeated, we should worry about other things. That is a fast number that we could put on a dashboard.

If the ratio is high enough to care about, we would then need to do the more expensive analysis of counting repeated signatures. If most repeats are 2-3 per day, then maybe we don't care. If we're getting 86,000 repeats a day, then it may be worth writing some C code. I thought some key collision would be insignificant.

But, this is optimizing before data. We could use 128-character signatures, then tune it down after an initial analysis.